crates/base/src/utils/graph_util.rs

```diff
@@ -0,0 +1,205 @@
+            ))
+        });
+    if let Some(error) = errors.next() {
+        Err(error)
+    } else {
+        Ok(())
+    }
+}
+
+pub async fn build_graph_with_npm_resolution<'a>(
+    graph: &mut ModuleGraph,
+    roots: Vec<ModuleSpecifier>,
+    loader: &mut dyn deno_graph::source::Loader,
+    options: deno_graph::BuildOptions<'a>,
+) -> Result<(), AnyError> {
+    //TODO: NPM resolvers
+
+    // ensure an "npm install" is done if the user has explicitly
+    // opted into using a node_modules directory
+    // if self.options.node_modules_dir_enablement() == Some(true) {
+    //     self.resolver.force_top_level_package_json_install().await?;
+    // }
+
+    graph.build(roots, loader, options).await;
+
+    // ensure that the top level package.json is installed if a
+    // specifier was matched in the package.json
+    // self
+    //     .resolver
+    //     .top_level_package_json_install_if_necessary()
+    //     .await?;
+
+    // resolve the dependencies of any pending dependencies
+    // that were inserted by building the graph
+    // self.npm_resolver.resolve_pending().await?;
+
+    Ok(())
+}
+
+pub async fn create_graph_and_maybe_check(
+    roots: Vec<ModuleSpecifier>,
+) -> Result<deno_graph::ModuleGraph, AnyError> {
+    let emitter_factory = EmitterFactory::new();
+
+    let mut cache = emitter_factory.file_fetcher_loader();
+    let analyzer = emitter_factory.parsed_source_cache().unwrap().as_analyzer();
+    let graph_kind = deno_graph::GraphKind::CodeOnly;
+    let mut graph = ModuleGraph::new(graph_kind);
+    let graph_resolver = emitter_factory.graph_resolver();
+
+    build_graph_with_npm_resolution(
+        &mut graph,
+        roots,
+        cache.as_mut(),
+        deno_graph::BuildOptions {
+            is_dynamic: false,
+            imports: vec![],
+            resolver: Some(&*graph_resolver),
+            npm_resolver: None,
+            module_analyzer: Some(&*analyzer),
+            reporter: None,
+        },
+    )
+    .await?;
+
+    //let graph = Arc::new(graph);
+    // graph_valid_with_cli_options(&graph, &graph.roots, &self.options)?;
+    // if let Some(lockfile) = &self.lockfile {
+    //     graph_lock_or_exit(&graph, &mut lockfile.lock());
+    // }
+
+    // if self.options.type_check_mode().is_true() {
+    //     self
+    //         .type_checker
+    //         .check(
+    //             graph.clone(),
+    //             check::CheckOptions {
+    //                 lib: self.options.ts_type_lib_window(),
+    //                 log_ignored_options: true,
+    //                 reload: true, // TODO: ?
+    //             },
+    //         )
+    //         .await?;
+    // }
+
+    Ok(graph)
+}
+
+pub async fn create_module_graph_from_path(
+    main_service_path: &str,
+) -> Result<ModuleGraph, Box<dyn std::error::Error>> {
+    let main_service_directory = PathBuf::from(main_service_path);
+    let index = main_service_directory.clone().join("index.ts");
```

Comment:
we can allow users to specify the full path to the entrypoint instead of the directory. That way we don't have to assume the entrypoint is always a `index.ts`